### PR TITLE
Adding provisioning documents

### PR DESCRIPTION
**Pull request by jebrando**

### CODE REVIEW

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Azure Iot Provisioning Client FAQ
+
+Here is where we shall try to answer question pertaining to the IoT Provisioning Client.  Please read the following documents for an overview of the DPS service.
+
```
Comment (yzhong94) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: I'd use "Device Provisioning Service" and link to our docs: https://docs.microsoft.com/en-us/azure/iot-dps/

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+## x509 Provisioning
+
+- How does Provisining match the x509 certificates to an enrollment?
+
+  - The Provisioning Client add the x509 certificate (more on the certificate in a second) on the TLS stream.  The DPS service uses this certificate to look up the proper enrollment record.
+
+- Where does the x509 certificate comes from?
+
+  - This question depends on the situation:
```
Comment (yzhong94) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: Make sure you link to: https://docs.microsoft.com/en-us/azure/iot-dps/concepts-security#x509-certificates

**Comment by yzhong94** [comment]
@nberdy for review as well

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# Azure Iot Provisioning Client FAQ
```
Comment (yzhong94) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: I want to see more!  It'd be great to have an architecture diagram to see how the two clients work together.  It'd also be great to have documentation around how the provisioning client and device client "should" interact.

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+Here is where we shall try to answer question pertaining to the IoT Provisioning Client.  Please read the following documents for an overview of the DPS service.
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: Don't say "try" even though it's a best-effort doc.

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+## x509 Provisioning
+
+- How does Provisining match the x509 certificates to an enrollment?
+
+  - The Provisioning Client add the x509 certificate (more on the certificate in a second) on the TLS stream.  The DPS service uses this certificate to look up the proper enrollment record.
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: There are two parts here: how the provisioning client uses the certificate, and how DPS actually matches the device to the right enrollment. Please see my email to Randy for a walkthrough + diagram for how we match to an enrollment (which the doc team also has and I thought would have turned into a doc)

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# Azure Iot Provisioning Client FAQ
+
+Here is where we shall try to answer question pertaining to the IoT Provisioning Client.  Please read the following documents for an overview of the DPS service.
+
+## x509 Provisioning
+
+- How does Provisining match the x509 certificates to an enrollment?
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: I'm going to be a grammar snob and edit your grammar all over, sorry in advance.

Should be: How does the provisioning service match the device's X.509 certificate to an enrollment?

File: provisioning_client/devdoc/azure_provisioning_faq.md (hunk 5 above)
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: "add" should be "adds"

Please move the parenthetical (more on the cert...) to the end of the sentence so it breaks it up less.

"The DPS service" should be "The device provisioning service uses..."

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+  - The Provisioning Client add the x509 certificate (more on the certificate in a second) on the TLS stream.  The DPS service uses this certificate to look up the proper enrollment record.
+
+- Where does the x509 certificate comes from?
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: s/comes/come

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+  - This question depends on the situation:
+
+    - **Development**: The SDK ships with a development repo that generates test x509 with a canned key.  This enables the developer to quickly get up and running to test their solutions.
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: s/canned/pre-generated

(Not all readers will be native English speakers)

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+    - **Production**: For productions situations the developer should create a custom HSM library to retrieve the certificate from a hardware backed HSM or a software solution (for more information on this please see [using custom hsm](https://github.com/Azure/azure-iot-sdk-c/blob/master/provisioning_client/devdoc/using_custom_hsm.md))
+
+- What if I want the certificate private key to never leave the hardware?
+
+  - One of the benifits of having an HSM is that it can ensure that the private key will never leave the hardware device, but to get this functionality it will require a little more code.
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: s/benifits/benefits

File: provisioning_client/devdoc/azure_provisioning_faq.md (hunk 5 above)
Comment (nberdy) [reply provisioning_client/devdoc/azure_provisioning_faq.md]: also you have "more on this in a moment" twice in the FAQ so far (second time is below) - is there any way to be a little more specific as to when it's coming up?

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+    - You will need to write a custom HSM to be able to extract the certificate and return the alias private key to the SDK (more on this in a moment) see the [custom hsm sample](https://github.com/Azure/azure-iot-sdk-c/blob/master/provisioning_client/samples/custom_hsm_example/custom_hsm_example.c) in the SDK.
+
+    - You will need to have a tls engine that can communicate with the target hardware that is connected to the device.
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: s/tls/TLS

Are there particular TLS engines we have tried and feel comfortable mentioning?

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    - You will also need to create a custom azure iot tlsio library to communicate with the hardware tls engine.  You can use the [tlsio_template](https://github.com/Azure/azure-c-shared-utility/blob/master/adapters/tlsio_template.c) to get started or you can look at an already complete tlsio such as [tlsio_openssl](https://github.com/Azure/azure-c-shared-utility/blob/master/adapters/tlsio_openssl.c)
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: When does a customer have to create their own engine vs use one that's pre-built? What are things to consider?

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+  - More on Alias Private Keys
+
+    - The custom HSM uses the concept of **Alias Private Keys** to deal with keeping the private keys information in the hardware.  The developer can pass a token that can be interpreted by the hardware stack and the SDK will simply pass the token down to the custom tlsio interface without inspecting the value of the token.  This way the actual keys never leave the hardware.
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: Is there a reference article that describes what an Alias Private Key is that you can link to?

Comment (nberdy) [reply provisioning_client/devdoc/azure_provisioning_faq.md]: Also, is there a diagram that could illustrate this?

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+## Symmetric Key Provisioning
+
+- Coming soon
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: Link to relevant doc: https://docs.microsoft.com/en-us/azure/iot-dps/concepts-symmetric-key-attestation

File: provisioning_client/devdoc/azure_provisioning_faq.md [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+## TPM Provisioning
+
+- Coming soon
```
Comment (nberdy) [inline provisioning_client/devdoc/azure_provisioning_faq.md]: Link to relevant doc: https://docs.microsoft.com/en-us/azure/iot-dps/concepts-tpm-attestation

File: provisioning_client/samples/readme.md [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+## List of samples
+
+- custom_hsm_examples
+
+  - This example shows how to create a custom HSM to retrieve the secruity information needed for the provisioning SDK to communication the the DPS Service.
+
+- iothub_client_sample_hsm
+
+  - This sample show how to use the IoThub Device Client with the device authentication system without using provisioning.
```
Comment (nberdy) [inline provisioning_client/samples/readme.md]: why is this in the provisioning_client folder if it doesn't use provisioning?

File: provisioning_client/samples/readme.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+- prov_dev_client_ll_sample
+
+  - This sample uses the non-threaded SDK to connect the provisioning client and get the IoThub information and then connects to the iothub.  This example would be for devices that require greater control over communications by calling DoWork to control the message pump.
```
Comment (nberdy) [inline provisioning_client/samples/readme.md]: s/iothub/IoT hub

File: provisioning_client/samples/readme.md [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+
+- prov_dev_client_sample
```
Comment (nberdy) [inline provisioning_client/samples/readme.md]: If I'm just getting started and don't have hardware yet (but I'm a C developer), which one of these samples do you recommend using? Or is it obvious to a real C developer?

File: provisioning_client/devdoc/azure_provisioning_faq.md (hunk 3 above)
Comment (jebrando) [reply provisioning_client/devdoc/azure_provisioning_faq.md]: Yes, we can do that.  Let's have a quick chat to talk about exactly what we do here.

---
In reply to: [254831673](https://github.com/Azure/azure-iot-sdk-c/pull/854#discussion_r254831673) [](ancestors = 254831673)

File: provisioning_client/devdoc/azure_provisioning_faq.md (hunk 11 above)
Comment (jebrando) [reply provisioning_client/devdoc/azure_provisioning_faq.md]: Great question.

---
In reply to: [254838049](https://github.com/Azure/azure-iot-sdk-c/pull/854#discussion_r254838049) [](ancestors = 254838049)

File: provisioning_client/samples/readme.md (hunk 15 above)
Comment (jebrando) [reply provisioning_client/samples/readme.md]: Good point.

---
In reply to: [254838897](https://github.com/Azure/azure-iot-sdk-c/pull/854#discussion_r254838897) [](ancestors = 254838897)

File: provisioning_client/devdoc/azure_provisioning_faq.md (hunk 12 above)
Comment (jebrando) [reply provisioning_client/devdoc/azure_provisioning_faq.md]: I would have to create a diagram.  I don't have anything yet.

---
In reply to: [254838585](https://github.com/Azure/azure-iot-sdk-c/pull/854#discussion_r254838585) [](ancestors = 254838585)

Comment (jebrando) [reply provisioning_client/devdoc/azure_provisioning_faq.md]: No this is something that I got from RIoT, they us Alias key concept, but it is never talked about.

---
In reply to: [254838373](https://github.com/Azure/azure-iot-sdk-c/pull/854#discussion_r254838373) [](ancestors = 254838373)

File: provisioning_client/devdoc/azure_provisioning_faq.md (hunk 11 above)
Comment (nberdy) [reply provisioning_client/devdoc/azure_provisioning_faq.md]: Would be great to explicitly state that - we'll have some hardware manufacturers reading our docs/looking at our SDKs who will want to know more.

Comment (massand) [reply provisioning_client/devdoc/azure_provisioning_faq.md]: I think we have sufficient info here from the openssl standpoint for hardware manufacturers to figure out they have to add to our PAL layer. The SDK provides the hooks for that and we have a great sample. More feedback / PRs from that community are welcome to help improve this.

**Pull request merged**